--[[
 		 _       _             _          _       _           _       
 		/ \   __| | ___  _ __ (_)___     / \   __| |_ __ ___ (_)_ __  
	   / _ \ / _` |/ _ \| '_ \| / __|   / _ \ / _` | '_ ` _ \| | '_ \ 
	  / ___ \ (_| | (_) | | | | \__ \  / ___ \ (_| | | | | | | | | | |
	 /_/   \_\__,_|\___/|_| |_|_|___/ /_/   \_\__,_|_| |_| |_|_|_| |_|
                                                                  

	    ______      _         ____                                             __           __
	   / ____/___  (_)  __   /  _/___  _________  _________  ____  _________ _/ /____  ____/ /
	  / __/ / __ \/ / |/_/   / // __ \/ ___/ __ \/ ___/ __ \/ __ \/ ___/ __ `/ __/ _ \/ __  / 
	 / /___/ /_/ / />  <   _/ // / / / /__/ /_/ / /  / /_/ / /_/ / /  / /_/ / /_/  __/ /_/ /  
	/_____/ .___/_/_/|_|  /___/_/ /_/\___/\____/_/  / .___/\____/_/   \__,_/\__/\___/\__,_/   
	     /_/                                       /_/                                        

	-----------------------
	-- Adonis MainModule --
	-----------------------

	Distributed under the MIT license.

	Please do not redistribute without proper credits.
--]]

return require(script.Server.Server)
