--[[
	 ______      __                                  ______      __                             
	/\  _  \    /\ \                  __            /\  _  \    /\ \              __            
	\ \ \L\ \   \_\ \    ___     ___ /\_\    ____   \ \ \L\ \   \_\ \    ___ ___ /\_\    ___    
	 \ \  __ \  /'_` \  / __`\ /' _ `\/\ \  /',__\   \ \  __ \  /'_` \ /' __` __`\/\ \ /' _ `\  
	  \ \ \/\ \/\ \L\ \/\ \L\ \/\ \/\ \ \ \/\__, `\   \ \ \/\ \/\ \L\ \/\ \/\ \/\ \ \ \/\ \/\ \ 
	   \ \_\ \_\ \___,_\ \____/\ \_\ \_\ \_\/\____/    \ \_\ \_\ \___,_\ \_\ \_\ \_\ \_\ \_\ \_\
	    \/_/\/_/\/__,_ /\/___/  \/_/\/_/\/_/\/___/      \/_/\/_/\/__,_ /\/_/\/_/\/_/\/_/\/_/\/_/

	  _____       _        ___                                            _           _ 
	 | ____|_ __ (_)_  __ |_ _|_ __   ___ ___  _ __ _ __   ___  _ __ __ _| |_ ___  __| |
	 |  _| | '_ \| \ \/ /  | || '_ \ / __/ _ \| '__| '_ \ / _ \| '__/ _` | __/ _ \/ _` |
	 | |___| |_) | |>  <   | || | | | (_| (_) | |  | |_) | (_) | | | (_| | ||  __/ (_| |
	 |_____| .__/|_/_/\_\ |___|_| |_|\___\___/|_|  | .__/ \___/|_|  \__,_|\__\___|\__,_|
	       |_|                                     |_|                                  

	-----------------------
	-- Adonis MainModule --
	-----------------------

	Distributed under the MIT license.

	Please do not redistribute without proper credits.
--]]

return require(script.Server.Server)
